tests/python/test_tools_smoke.py

```diff
@@ -0,0 +1,371 @@
+#!/usr/bin/env python
+# Copyright (c) Sasha Goldshtein, 2017
+# Licensed under the Apache License, Version 2.0 (the "License")
+
+import distutils.version
+import subprocess
+import os
+from unittest import main, skipUnless, TestCase
+
+TOOLS_DIR = "../../tools/"
+
+def kernel_version_ge(major, minor):
+    # True if running kernel is >= X.Y
+    version = distutils.version.LooseVersion(os.uname()[2]).version
+    if version[0] > major:
+        return True
+    if version[0] < major:
+        return False
+    if minor and version[1] < minor:
+        return False
+    return True
+
+@skipUnless(kernel_version_ge(4,1), "requires kernel >= 4.1")
+class SmokeTests(TestCase):
+    # Use this for commands that have a built-in timeout, so they only need
+    # to be killed in case of a hard hang.
+    def run_with_duration(self, command, timeout=10):
+        full_command = TOOLS_DIR + command
+        self.assertEqual(0,     # clean exit
+                subprocess.call("timeout -s KILL %ds %s > /dev/null" %
+                                (timeout, full_command), shell=True))
+
+    # Use this for commands that don't have a built-in timeout, so we have
+    # to Ctrl-C out of them by sending SIGINT. If that still doesn't stop
+    # them, send a kill signal 5 seconds later.
+    def run_with_int(self, command, timeout=5, allow_early=False, kill=False):
+        full_command = TOOLS_DIR + command
+        signal = "KILL" if kill else "INT"
+        rc = subprocess.call("timeout -s %s -k %ds %ds %s > /dev/null" %
+                             (signal, 5, timeout, full_command), shell=True)
+        # timeout returns 124 if the program did not terminate prematurely,
+        # and returns 137 if we used KILL instead of INT. So there are three
+        # sensible scenarios:
+        #   1. The script is allowed to return early, and it did, with a
+        #      success return code.
+        #   2. The script timed out and was killed by the SIGINT signal.
+        #   3. The script timed out and was killed by the SIGKILL signal, and
+        #      this was what we asked for using kill=True.
+        self.assertTrue((rc == 0 and allow_early) or rc == 124
+                        or (rc == 137 and kill), "rc was %d" % rc)
+
+    def setUp(self):
+        pass
+
+    def tearDown(self):
+        pass
+
+    def test_argdist(self):
+        self.run_with_duration("argdist.py -C 'p::SyS_open()' -n 1 -i 1")
+
+    @skipUnless(kernel_version_ge(4,4), "requires kernel >= 4.4")
+    def test_bashreadline(self):
+        self.run_with_int("bashreadline.py")
```

Comment:
I don't know if it's necessary. What we're doing here is almost identical: launch the tool, give it time to initialize, and then exit. It would be a lot of work adding this to all the tools, and I'm not sure what we'd gain.